in-progress/proving-queue/0005-proving-queue.md

```diff
@@ -0,0 +1,97 @@
+
+![Proving Architecture](./proving-arch.png)
+
+### A Simple but Sufficient Prover Implementation
+
+Frankly, there are many ways in which the `Prover` component can be built. It is a common pattern within Web 2 and many technologies exist to help. Ideally however we would like to reduce our dependencies as much as possible, so a solution relying solely on existing dependencies would be desirable. First, let us look at our high level requirements:
+
+1. Proving a single block of 11520 transactions (10 TPS for 32 blocks where each block is 3 ethereum blocks)
+2. 3 public functions per transaction results in the other of 100,000 proof requests.
+3. Depending on timeliness requirements and proof computational complexity, we should expect the number of prover agents to be up to 10,000.
+
+### Other Possibilities Considered
+
+We considered using RabbitMQ. This is a well supported persisted message queue implementation. However, like most queueing implementations, it's priorities are related to resilience, scalability and capacity. It can distribute a large number of jobs across a large number of clients very quickly. It also offers redundant setups etc. Also like most queueing implementations, it lacks flexibility. For example, it's possible that proving agents won't be homogenous. A prover may have access to a variety of machines with differing hardware capabilities meaning greater flexibility is needed when it comes to retrieving jobs and distributing them. This type of flexibility isn't generally supported by off the shelf queueing implementations.
+
+### LMDB Backed Proof Request Store
+
+We propose creating a proving job broker as a Node.js server using an LMDB database as it's persistent store. Crucially, we will make a number of compromises to achieve this.
+
+#### Broker Architecture and Implementation
+
+The broker exists as an HTTP service using an LMDB DB to store an index of proving jobs. The index is a mapping from the job id to the job's metadata. The job metadata can be summarised by the following:
+
+``` TS
+type Metadata = {
+  proofType: PROOF_TYPE; //(BASE_ROLLUP etc...)
+  epochNumber: number;
+  status: // (IN_PROGRESS or COMPLETED)
+}
+
+```
+
+Proving request data, such as input witness and recursive proofs are stored in a directory labelled with the job's id residing on an NFS share/S3 bucket or similar. This is an optimsation, as prover agents can access the data independently, without requiring the broker to transfer large amounts of data to them. If it turns out that this is not required then the proof requests will reside on a disk local to the broker and will be transferred over the network from the broker. Maybe this should be a configurable aspect of the system.
```

Comment:
Unless we see the amount of data being too large, I'd leave this optimization out, since having a shared NFS may not be an option for decentralized proving marketplaces, and an S3 bucket ties the implementation to AWS. I suspect the amount of data we need to push is not _too_ much.

Reply:
Actually, we can probably count on an S3 bucket anyway. It is only coupled to the prover-node, from the prover-agent's perspective it's just a URL that they use for downloading proving work.